### PR TITLE
Make Fix-All use normal text-diffing as the TreeDiffer produces some very bad results.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide 
Note: tree diffing issue is reported here: https://github.com/dotnet/roslyn/issues/17498

**Comment by CyrusNajmabadi** [comment]
retest windows_debug_unit32_prtest please

**Comment by CyrusNajmabadi** [comment]
retest ubuntu_14_debug_prtest please

**Comment by CyrusNajmabadi** [comment]
retest windows_debug_vs-integration_prtest please

**Comment by CyrusNajmabadi** [comment]
retest windows_release_vs-integration_prtest please

**Comment by CyrusNajmabadi** [comment]
retest ubuntu_16_debug_prtest please

File: src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/BatchFixAllProvider.cs [hunk 1]
```diff
@@ -371,80 +376,182 @@ public virtual string GetFixAllTitle(FixAllState fixAllState)
...
 
         private static bool AllChangesCanBeApplied(
-            SimpleIntervalTree<TextChange> cumulativeChanges, IList<TextChange> currentChanges)
+            SimpleIntervalTree<TextChange> cumulativeChanges,
+            ImmutableArray<TextChange> currentChanges)
         {
             var overlappingSpans = ArrayBuilder<TextChange>.GetInstance();
+            var intersectingSpans = ArrayBuilder<TextChange>.GetInstance();
+
+            var result = AllChangesCanBeApplied(
```
Comment (dpoeschl) [inline src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/BatchFixAllProvider.cs]: I don't like calling things "result", even if it is very temporary.

File: src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/BatchFixAllProvider.cs [hunk 2]
```diff
@@ -371,80 +376,182 @@ public virtual string GetFixAllTitle(FixAllState fixAllState)
...
-                    }
+                cumulativeChanges.FillWithIntervalsThatIntersectWith(
+                   change.Span.Start, change.Span.Length, intersectingSpans);
 
-                    // Changes weren't the same.  Can't merge this in.
-                    return false;
-                }
-                else
+                var value = ChangeCanBeApplied(change,
+                    overlappingSpans: overlappingSpans,
```
Comment (dpoeschl) [inline src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/BatchFixAllProvider.cs]: No named args needed.

File: src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/BatchFixAllProvider.cs [hunk 3]
```diff
@@ -371,80 +376,182 @@ public virtual string GetFixAllTitle(FixAllState fixAllState)
...
             return true;
         }
+
+        private static bool ChangeCanBeApplied(
+            TextChange change,
+            ArrayBuilder<TextChange> overlappingSpans,
+            ArrayBuilder<TextChange> intersectingSpans)
+        {
+            // Pure insertions (i.e. changes with empty spans) are complex to evaluate.  
+            // We have to make sure that
```
Comment (dpoeschl) [inline src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/BatchFixAllProvider.cs]: Don't leave me hanging!

File: src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/BatchFixAllProvider.cs [hunk 4]
```diff
@@ -371,80 +376,182 @@ public virtual string GetFixAllTitle(FixAllState fixAllState)
...
             // All the changes would merge in fine.  We can absorb this.
             return true;
         }
+
+        private static bool ChangeCanBeApplied(
+            TextChange change,
+            ArrayBuilder<TextChange> overlappingSpans,
+            ArrayBuilder<TextChange> intersectingSpans)
+        {
+            // Pure insertions (i.e. changes with empty spans) are complex to evaluate.
```
Comment (dpoeschl) [inline src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/BatchFixAllProvider.cs]: "changes with empty overlapping/intersecting spans" or some other wording?

File: src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/BatchFixAllProvider.cs [hunk 5]
```diff
@@ -371,80 +376,182 @@ public virtual string GetFixAllTitle(FixAllState fixAllState)
...
+            ArrayBuilder<TextChange> intersectingSpans)
+        {
+            Debug.Assert(!change.Span.IsEmpty);
+
+            // We care about our intersections with *empty-span* changes.  Non-empty changes that
+            // we overlap are already handled in OverwriteChangeConflictsWithOverlappingSpans.
+            // And non-empty spans that we abut (i.e. which we're adjacent to) are totally safe 
+            // for both to be applied.
+            //
+            // However, empty-span changes are extremely ambiguous. It is not possible to tell which
```
Comment (dpoeschl) [inline src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/BatchFixAllProvider.cs]: "empty-span changes" is a weird term. Can you just better define "pure insertion" at the beginning and then rely on that term instead?

**Pull request merged**